src/odoc.ml

```diff
@@ -263,329 +267,364 @@ module Gen (S : sig val sctx : SC.t end) = struct
+        , Lib.hash l
+        , Or_exn.hash (List.hash Lib.hash) r
+        )
 
-  let entry_modules_by_lib lib =
-    Dir_contents.get sctx ~dir:(Lib.src_dir lib)
-    |> Dir_contents.modules_of_library ~name:(Lib.name lib)
-    |> Lib_modules.entry_modules
-
-  let entry_modules ~pkg =
-    libs_of_pkg ~pkg
-    |> Lib.Set.to_list
-    |> List.filter_map ~f:(fun l ->
-      if Lib.is_local l then (
-        Some (l, entry_modules_by_lib l)
-      ) else (
-        None
-      ))
-    |> Lib.Map.of_list_exn
-
-  let default_index ~pkg entry_modules =
-    let b = Buffer.create 512 in
-    Printf.bprintf b "{0 %s index}\n"
-      (Package.Name.to_string pkg);
-    Lib.Map.to_list entry_modules
-    |> List.sort ~compare:(fun (x, _) (y, _) ->
-      Lib_name.compare (Lib.name x) (Lib.name y))
-    |> List.iter ~f:(fun (lib, modules) ->
-      Printf.bprintf b "{1 Library %s}\n" (Lib_name.to_string (Lib.name lib));
-      Buffer.add_string b (
-        match modules with
-        | [ x ] ->
-          sprintf
-            "The entry point of this library is the module:\n{!module-%s}.\n"
-            (Module.Name.to_string (Module.name x))
-        | _ ->
-          sprintf
-            "This library exposes the following toplevel modules:\n\
-             {!modules:%s}\n"
-            (modules
-             |> List.sort ~compare:(fun x y ->
-               Module.Name.compare (Module.name x) (Module.name y))
-             |> List.map ~f:(fun m -> Module.Name.to_string (Module.name m))
-             |> String.concat ~sep:" ")
-      );
-    );
-    Buffer.contents b
-
-  let check_mlds_no_dupes ~pkg ~mlds =
-    match
-      List.map mlds ~f:(fun mld ->
-        (Filename.chop_extension (Path.basename mld), mld))
-      |> String.Map.of_list
-    with
-    | Ok m -> m
-    | Error (_, p1, p2) ->
-      die "Package %s has two mld's with the same basename %s, %s"
-        (Package.Name.to_string pkg)
-        (Path.to_string_maybe_quoted p1)
-        (Path.to_string_maybe_quoted p2)
-
-  let setup_package_odoc_rules_def =
-    let module Input = struct
-      type t = Package.Name.t * Path.t list
-
-      let hash (p, ps) =
-        Hashtbl.hash (Package.Name.hash p, List.hash Path.hash ps)
-
-      let equal (x1, y1) (x2, y2) =
-        Package.Name.equal x1 x2 && List.equal Path.equal y1 y2
-
-      let to_sexp (name, paths) =
-        Dyn.Tuple
-          [ Package.Name.to_dyn name
-          ; Dyn.List (List.map ~f:Path.to_dyn paths)
-          ]
-        |> Dyn.to_sexp
-    end
-    in
-    Memo.create "setup-package-odoc-rules"
-      ~output:(Allow_cutoff (module Unit))
-      ~doc:"setup odoc package rules"
-      ~input:(module Input)
-      ~visibility:Hidden
-      Sync
-      None
+    let to_sexp _ = Sexp.Encoder.string "<opaque>"
+  end
+  in
+  Memo.create "setup-library-html-rules"
+    ~doc:"setup html rules for library"
+    ~input:(module Input)
+    ~output:(Allow_cutoff (module Unit))
+    ~visibility:Hidden
+    Sync
+    None
```

Comment:
Note that we could provide the implementation immediately here

Reply:
Yeah, I mostly just like doing it later as it's easier to avoid accidentally capturing stuff in the closure.